src/Microsoft.AspNetCore.Routing.Abstractions/RouteValueDictionary.cs

```diff
@@ -35,62 +38,51 @@ public RouteValueDictionary()
         {
-            var otherDictionary = values as RouteValueDictionary;
-            if (otherDictionary != null)
+            RouteValueDictionary dictionary;
+            IEnumerable<KeyValuePair<string, object>> keyValueEnumerable;
+            if ((dictionary = values as RouteValueDictionary) != null)
             {
-                if (otherDictionary.InnerDictionary != null)
-                {
-                    InnerDictionary = new Dictionary<string, object>(
-                        otherDictionary.InnerDictionary.Count,
-                        StringComparer.OrdinalIgnoreCase);
-
-                    foreach (var kvp in otherDictionary.InnerDictionary)
-                    {
-                        InnerDictionary[kvp.Key] = kvp.Value;
-                    }
-
-                    return;
-                }
-                else if (otherDictionary.Properties != null)
+                ListStorage listStorage;
+                if ((listStorage = dictionary._storage as ListStorage) != null)
                 {
-                    Properties = otherDictionary.Properties;
-                    Value = otherDictionary.Value;
+                    _storage = new ListStorage(listStorage);
                     return;
                 }
-                else
+
+                PropertyStorage propertyStorage;
+                if ((propertyStorage = dictionary._storage as PropertyStorage) != null)
                 {
+                    _storage = new PropertyStorage(propertyStorage);
                     return;
                 }
-            }
 
-            var keyValuePairCollection = values as IEnumerable<KeyValuePair<string, object>>;
-            if (keyValuePairCollection != null)
+                // If we get here, it's an EmptyStorage
+                _storage = EmptyStorage.Instance;
+                return;
+            }
+            else if ((keyValueEnumerable = values as IEnumerable<KeyValuePair<string, object>>) != null)
             {
-                InnerDictionary = new Dictionary<string, object>(StringComparer.OrdinalIgnoreCase);
-
-                foreach (var kvp in keyValuePairCollection)
+                _storage = new ListStorage();
+                foreach (var kvp in keyValueEnumerable)
                 {
-                    InnerDictionary[kvp.Key] = kvp.Value;
+                    // TrySetValue always succeeds for ListStorage
+                    _storage.TrySetValue(kvp.Key, kvp.Value);
                 }
 
                 return;
             }
-
-            if (values != null)
+            else if (values != null)
             {
-                Properties = PropertyHelper.GetVisibleProperties(values);
-                Value = values;
-
+                _storage = new PropertyStorage(values, PropertyHelper.GetVisibleProperties(values));
+                return;
+            }
+            else
+            {
+                _storage = EmptyStorage.Instance;
                 return;
```

Comment:
Most of the `return` statements in this constructor are just noise and make things longer. Either use early exits everywhere (removing the `else if` bits) or remove excess `return`s.